x-pack/plugins/lists/server/services/exception_lists/exception_list_client.ts

```diff
@@ -46,10 +46,13 @@ import { findExceptionListItem } from './find_exception_list_item';
 import { findExceptionList } from './find_exception_list';
 import { findExceptionListsItem } from './find_exception_list_items';
 import { createEndpointList } from './create_endpoint_list';
+import { createEndpointTrustedAppsList } from './create_endpoint_trusted_apps_list';
 
 export class ExceptionListClient {
   private readonly user: string;
 
+  private trustedAppsListCreated: boolean = false;
```

Comment:
Added this for efficiency - see further below

Reply:
If the trusted apps list gets deleted this boolean will prevent you from recreating the list. We also need to be careful about relying on the boolean since it won't be shared across multiple kibana instances.

Reply:
Why would it be deleted?

Also with multiple instance, then each instance does this at most once, which is ok right?

Reply:
This is probably benign, but could be a little dangerous... this effectively makes the `createdTrustedAppsList` method not concurrency-safe. While we know that it's only called on plugin startup once per instance, this functionality is available as a public method on a client that can be called by anyone at any time... I think it would be better to check for existence of the list, and then create it if it's not there (from the plugin startup, since we know that it will only be called once).

Reply:
@madirey summed it up well, adding a boolean here creates state in each kibana instance that could get out of sync with the true state in ES - I'm hesitant to add local state for a small optimization unless we really need it.

Reply:
Thank you both for the feedback.

Personally I don't see the purpose of keep calling ES (through saved object client) to attempt to create the list every time you`createEndpointListItem()` or `updateEndpointListItem()` or `findEndpointListItem`. This state property avoids that for trusted apps only.

Re:
> While we know that it's only called on plugin startup once per instance, this functionality is available as a public method on a client that can be called by anyone at any time...

I also don't see an issue here. It actually would behave the same as the code you have in place today minus actually attempting to create the list. `createEndpointTrustedAppsList()`behaves exactly the same way as `createEndpointList()` service  (returns `null` instead of `ExceptionListSchema` if list already exists).

That being said - I can make a change. So are you suggesting that I remove this internal state from the Class (I think that is the only thing needed - correct?)

Reply:
I would remove the internal state but not add calls to the `createEndpointTrustedAppsList()` method in other areas. I think we should be ok with returning a `list not found` error on the `findItem`, `updateItem`, and `createItem` APIs if the list doesn't exist and leave it up to users or the application to re-create the list if desired. I realize that isn't how the endpoint list APIs work right now so we should consider changing those to match in a separate PR.

Reply:
Yeah, with something like this there is another small state issue which is...
 
With the way this is written it looks like each plugin instance will get by default a `trustedAppsListCreated` that is initially set to be false. So if for some reason multiple list plugin instances down the road in the same Kibana container are trying to initialize your trusted apps list, they will each need to make an attempt at different times and to have this flag flipped to be true. If this was a static member instead:

```ts
 private static trustedAppsListCreated: boolean = false;
```

Then you don't have the worry of multiple instances of this class having different views of `trustedAppsListCreated` where one instance of this class thinks it's true and another instance actually thinks it's false because it never called into it trying to create it. Not a big deal, but it just shows some of the perils of objects and states that are easy to overlook.

I'm not personally a big "OOP person/class based person" compared to using small pure functions as the pure small functions always seem to be better in a lot of ways. This class is mostly here because of the Kibana methodology of using classes for plugins when doing plugin work and I tried to keep state management to be immutable and use the underlying elastic data store as our source of truth rather than caching things. Even things as seemingly trivial as caching that something does exist or does not exist will eventually have issues so we are very careful about using state. 

>Why would it be deleted?

Users with access to the Kibana dev tools, saved object API's, etc... do unexpected things at different times. Although we don't expose these types of saved objects to the UI of the kibana SO management screens here:

```ts
${YOUR_URL}/app/management/kibana/objects
```

At some point we might end up exposing them there or write other user interface screens which do allow deletions. But, fwiw, nothing is stopping users from deleting these objects using the saved object API with some REST/CURL. We have seen users delete parts of their saved objects during runtime before as an example or just delete subsets of pieces of data they didn't want using elastic search directly. 

If a user does delete this piece of data, they will have to restart at least one Kibana instance to re-initialize this structure and the others will be more than likely sitting at a "true" state. Not too terrible, but those Kibana instances will have a false view of this being created when it's not.

The last piece of thing with this small piece of state is if during startup your Kibana instance gets an error trying to create this structure then it will keep this flag set to false until the next time it does a retry (if it doesn't do a retry it is forever false until restarted). Not a big deal now, but later if you're relying on that flag for other things it would start to become an issue.

>I think we should be ok with returning a list not found error on the findItem, updateItem, and createItem APIs if the list doesn't exist and leave it up to users or the application to re-create the list if desired.

We can move back that direction. We had it kind of that way to start with but then the UI consumers were always needing to call within the UI checks to see if it existed or not before calling those API's so we optimized them to auto-call things. However, if we want them to just make a call and if they get a 404, then they have to provision them, we can.


As for optimizations go and communication concerns around Kibana with Elastic and how far apart they really are. This is a good guide:
https://www.elastic.co/guide/en/kibana/current/production.html

```
If you have multiple nodes in your Elasticsearch cluster, the easiest way to distribute
Kibana requests across the nodes is to run an Elasticsearch Coordinating only node
on the same machine as Kibana. Elasticsearch Coordinating only nodes are essentially
smart load balancers that are part of the cluster. They process incoming HTTP requests,
redirect operations to the other nodes in the cluster as needed, and gather and return
the results. For more information, see Node in the Elasticsearch reference.
```

Typically I think that Kibana always has a local coordinating Elasticsearch so REST between the two should be fairly quick if anyone is worried about that.

Reply:
I will be reverting this to follow the same current implementation at Exceptions (each of the mentioned methods will "auto-create" the list if not already there).

Thanks for the feedback.